RFC for Boxes

SamFerracin opened the pull request.

SamFerracin committed:
first pass

SamFerracin committed:
some

SamFerracin committed:
changed some wording

ihincks commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,185 @@
+| **Submitted**     | YYYY-MM-DD                                   |
+
+## Summary
+Objective: Introducing a new ``CircuitBlock`` class to enhance the user experience with Qiskit's built-in primitives, with the goal of providing more flexibility and full transparency with respect to twirling and mitigation.
```
```suggestion
Objective: Introduce a new ``CircuitBlock`` instruction to enhance the user experience with Qiskit's Runtime primitives, with the goal of providing more flexibility and transparency with respect to twirling and mitigation.
```

ihincks replied:
Separately, I might just call it a `Block`. 

Also, incidentally, this is markdown so you want single backticks.

ihincks commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,185 @@
+## Summary
+Objective: Introducing a new ``CircuitBlock`` class to enhance the user experience with Qiskit's built-in primitives, with the goal of providing more flexibility and full transparency with respect to twirling and mitigation.
+
+The concept of a "block" of gates—meaning a subset of gates isolated from the rest of the circuit—is fundamental to many quantum computing routines. However, Qiskit does not provide a dedicated object to represent a block. Traditionally, users have worked around this by defining blocks *indirectly* using barriers. While this approach has been adequate for tasks like scheduling and transpilation,  ...
```
I have some words that might somehow be edited into this paragraph:

A block of gates is a collection of circuit instructions that is promised to have a consistent execution when applied to a particular set of qubits. For example, the relative scheduling of gates within a block is consistent between multiple applications of the block. A circuit layer, or cycle, is a special case of a block that is full-width and depth-1.

In particular, a block is a natural concept when reasoning about quantum errors because the error profile often depends on factors such as which gates are applied simultaneously.

SamFerracin replied:
I stole the second sentence. Feel free to add the first one too if you find a natural spot for it :)

ihincks commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,185 @@
+
+The concept of a "block" of gates—meaning a subset of gates isolated from the rest of the circuit—is fundamental to many quantum computing routines. However, Qiskit does not provide a dedicated object to represent a block. Traditionally, users have worked around this by defining blocks *indirectly* using barriers. While this approach has been adequate for tasks like scheduling and transpilation,  ...
+
+The overcome these issues, we propose to establish the concept of a block of gates through the introduction of a new ``CircuitBlock`` class. The primary goal of this class is to encapsulate an isolated block of ``CircuitInstruction``s, treating it as a single unit for tasks like twirling and mitigation. We believe that ``CircuitBlock``s will offer greater transparency in pre-processing routines f ...
```
```suggestion
To overcome these issues, we propose to establish the concept of a block of gates through the introduction of a new ``CircuitBlock`` instruction class. The primary goal of this class is to encapsulate an isolated block of ``CircuitInstruction``s, treating it as a single unit for tasks like twirling and mitigation. We believe that ``CircuitBlock``s will offer greater transparency in pre-processing routines for twirling and mitigation, increased flexibility in how circuits are twirled or mitigated, and improved runtimes for certain mitigation experiments.
```

ihincks commented on 0022-circuit-block.md:
````diff
@@ -0,0 +1,185 @@
+q_1: ─░─┤ X ├─░─┤ X ├─░─┤ X ├─░─┤ X ├─░─┤ X ├─░─┤ X ├─░─
+      ░ └───┘ ░ └───┘ ░ └───┘ ░ └───┘ ░ └───┘ ░ └───┘ ░ 
+```
+To perform the noise-learning part of the protocol, the mitigation algorithm needs to break this circuit into blocks, and in order to do so, it can choose to handle barriers in three different ways:
````
```suggestion
To perform the noise-learning part of the protocol, the mitigation algorithm needs to break this circuit into blocks, and in order to do so, it might choose to handle barriers in three different ways:
```

ihincks commented on 0022-circuit-block.md:
````diff
@@ -0,0 +1,185 @@
+```
+To perform the noise-learning part of the protocol, the mitigation algorithm needs to break this circuit into blocks, and in order to do so, it can choose to handle barriers in three different ways:
+
+1. Including the barriers into the blocks (as is currently done).
+2. Ignoring the barriers.
+3. Switching to a new blocks when a barrier is encountered.
````
```suggestion
1. Include the barriers into the blocks (as is currently done).
2. Ignore the barriers.
3. Switch to a new block when a barrier is encountered.
```

ihincks commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,185 @@
+
+The overcome these issues, we propose to establish the concept of a block of gates through the introduction of a new ``CircuitBlock`` class. The primary goal of this class is to encapsulate an isolated block of ``CircuitInstruction``s, treating it as a single unit for tasks like twirling and mitigation. We believe that ``CircuitBlock``s will offer greater transparency in pre-processing routines f ...
+
+## Motivation
```
Reading this motivation section, I am expecting that it will come across as unmotivating for people not intimately familiar with the idiosyncratic issues of block collection in pec-runtime, that is, just about all non-author readers. 

I think we could write a stronger motivation section if we focused on the idea that a user won't know how we collect blocks until they get results, rather than this particular enumeration. I expect the common response to this section will be "duh, 3 is obviously best; do greedy collection"

ihincks commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,185 @@
+- All of the existing transpiler passes can be asked (at least initially) to simply ignore the blocks, since transpilation is meant to happen before the blocks are generated. This can save quite a lot of time that would otherwise be spent adding logic to every single transpiler pass -- and yet, we can choose to do this in the future.
+- Users of the primitives can inspect the blocks before submitting their jobs. This is a big improvement: today, the block generation happens in the server and can be unintuitive (see the example in the background section), and users have to wait for the job to be over before they can find out how the circuit was broken into blocks.
+
+In addition to supportin the two workflow above, we believe that the following workflow should not be supported:
```
```suggestion
In addition to supportin the two workflow above, we believe that the following workflow should *not* be supported:
```

ihincks commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,185 @@
+    - They submit their job.
+    - The server creates blocks for the users "behind the scenes."
+
+This workflow essentially takes the same steps as are taken today and therefore presents the same disadvantages as we have today, thus we believe that is should be banned.
```
```suggestion
This workflow essentially takes the same steps as are taken today and therefore presents the same disadvantages discussed above, and also violates the principle "don't do hidden transpilation server-side".
```

SamFerracin committed:
Update 0022-circuit-block.md

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

ihincks commented on 0022-circuit-block.md:
````diff
@@ -0,0 +1,185 @@
+```
+
+## Outstanding questions:
+1. In the existing implementation of blocks that exists in `pec-runtime`, blocks are hashable, so that they can be used as keys in dictionaries (specifically in the dictionary porduced by the ``NoiseLearner`` that maps blocks to noise models). This is clearly unsafe since circuits are mutable. Can we find a way around this? 
````
In particular, how are users going to tie noise models to blocks if we can't have a `dict[CircuitInstruction[Block], NoiseModel]` or equivalent. This is a crucial feature, so we need to have this planned for.

SamFerracin committed:
md

ihincks commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,185 @@
+## Outstanding questions:
+1. In the existing implementation of blocks that exists in `pec-runtime`, blocks are hashable, so that they can be used as keys in dictionaries (specifically in the dictionary porduced by the ``NoiseLearner`` that maps blocks to noise models). This is clearly unsafe since circuits are mutable. Can we find a way around this? 
+2. Should ``CircuitBlock`` added to qiskit core, alongside the transpiler pass groups gates in blocks workflow 1? If so, should they be written in Rust or Python?
+3. What packages and subpackages need to be modify once ``CircuitBlocks`` are introduced? For example, should we add logic to qiskit-aer to be able to simulate circuits with blocks, or to Qiskit's optimizer, ..?
```
```suggestion
3. What packages and subpackages need to be modify once ``CircuitBlocks`` are introduced? For example, should we add logic to qiskit-aer to be able to simulate circuits with blocks, or to Qiskit's optimizer, ..?
4. How can we (should we?) facilitate re-use of CircuitInstruction or Block instances? For example, suppose we want the same block in a circuit 100 times:
```python
for _ in range(100):
    with circuit.block(context=my_context, name="foo") as block:
        block.h(0) 
        block.cx(0, 1) 
```

SamFerracin committed:
Update 0022-circuit-block.md

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

SamFerracin committed:
Update 0022-circuit-block.md

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

SamFerracin committed:
Update 0022-circuit-block.md

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

SamFerracin committed:
Update 0022-circuit-block.md

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

SamFerracin committed:
Update 0022-circuit-block.md

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

SamFerracin committed:
CR

SamFerracin committed:
conflicts

SamFerracin committed:
Update 0022-circuit-block.md

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

SamFerracin committed:
CR

SamFerracin committed:
done with CR

joshuasn commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,152 @@
+## Summary
+Objective: Introduce a new `Block` instruction to enhance the user experience with Qiskit's Runtime primitives, with the goal of providing more flexibility and transparency with respect to twirling and mitigation.
+
+The concept of a block of gates is fundamental to many quantum computing routines. For example, a block is a natural concept when reasoning about quantum errors because the error profile often depends on factors such as which gates are applied simultaneously. However, Qiskit does not provide a dedicated object to represent a block. Traditionally, users have worked around this by defining blocks * ...
```
```suggestion
The concept of a block of gates is fundamental to many quantum computing routines. For example, a block is a natural concept when reasoning about quantum errors because the error profile often depends on factors such as which gates are applied simultaneously. However, Qiskit does not provide a dedicated object to represent a block. Traditionally, users have worked around this by defining blocks *indirectly* using barriers. While this approach has been adequate for tasks like scheduling and transpilation, barriers can be hard to parse when pre-processing circuits in preparation for twirling or mitigation. Suboptimal handling of barriers in these post-processing steps can introduce significant slowdowns in some mitigation experiments (for example, it can unnecessarily increase the time required by the noise learning steps) and produce outcomes that may seem unintuitive to typical users.
```

joshuasn commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,152 @@
+
+The concept of a block of gates is fundamental to many quantum computing routines. For example, a block is a natural concept when reasoning about quantum errors because the error profile often depends on factors such as which gates are applied simultaneously. However, Qiskit does not provide a dedicated object to represent a block. Traditionally, users have worked around this by defining blocks * ...
+
+The overcome these issues, we propose to establish the concept of a block of gates through the introduction of a new `Block` class. The primary goal of this class is to encapsulate an isolated block of `CircuitInstruction`s, treating it as a single unit for tasks like twirling and mitigation. We believe that `Block`s will offer:
```
```suggestion
To overcome these issues, we propose to establish the concept of a block of gates through the introduction of a new `Block` class. The primary goal of this class is to encapsulate an isolated block of `CircuitInstruction`s, treating it as a single unit for tasks like twirling and mitigation. We believe that `Block`s will offer:
```

SamFerracin committed:
Update 0022-circuit-block.md

Co-authored-by: joshuasn <53916441+joshuasn@users.noreply.github.com>

SamFerracin committed:
Update 0022-circuit-block.md

Co-authored-by: joshuasn <53916441+joshuasn@users.noreply.github.com>

joshuasn commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,152 @@
+    - They use (a convenience method built around) a new transpiler pass that collects circuit's gates into blocks, with the ability of specifying different collection strategies.
+    - They submit their job.
+2. The workflow for power users:
+    - Users initialize a `QuantumCircuit` adding blocks manually as they wish.
```
What does `as they wish` mean specifically? For a regular user, the workflow implies the blocks get added after the circuit is mapped to an ISA circuit. Is the idea that a power user could specify the circuit and blocks at the ISA level or block the circuit before the mapping?

SamFerracin replied:
Yes, exactly.
It's not hard to directly write a circuit that complies directly with ISA requirements, you just need to use the right gates and have the correct coupling map in front of you. Power users can do this, or also, apply their own passes to the sub-circuits and then inserting them into the bigger circuit as blocks

joshuasn commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,152 @@
+
+## Outstanding questions:
+1. In the existing implementation of blocks that exists in `pec-runtime`, blocks are hashable, so that they can be used as keys in dictionaries (specifically in the dictionary porduced by the `NoiseLearner` that maps blocks to noise models). This is clearly unsafe since circuits are mutable. How can we design the noise handle so that it is safer than that, but also more flexible (e.g., if users w ...
+2. Should `Block` added to qiskit core, alongside the transpiler pass groups gates in blocks workflow 1? If so, should they be written in Rust or Python?
```
```suggestion
2. Should `Block` be added to qiskit core, alongside the transpiler pass groups gates in blocks workflow 1? If so, should they be written in Rust or Python?
```

joshuasn commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,152 @@
+## Outstanding questions:
+1. In the existing implementation of blocks that exists in `pec-runtime`, blocks are hashable, so that they can be used as keys in dictionaries (specifically in the dictionary porduced by the `NoiseLearner` that maps blocks to noise models). This is clearly unsafe since circuits are mutable. How can we design the noise handle so that it is safer than that, but also more flexible (e.g., if users w ...
+2. Should `Block` added to qiskit core, alongside the transpiler pass groups gates in blocks workflow 1? If so, should they be written in Rust or Python?
+3. What packages and subpackages need to be modify once `Block` are introduced? For example, should we add logic to qiskit-aer to be able to simulate circuits with blocks, or to Qiskit's optimizer, ..?
```
```suggestion
3. What packages and subpackages need to be modified once `Block` is introduced? For example, should we add logic to qiskit-aer to be able to simulate circuits with blocks, or to Qiskit's optimizer, ..?
```

joshuasn commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,152 @@
+| **Status**        | **Proposed** |
+|:------------------|:---------------------------------------------|
+| **RFC #**         | 0022                                         |
+| **Authors**       | [Sam Ferracin](sam.ferracin@ibm.com), [Ian Hincks](ian.hincks@ibm.com), [Chris Wood](cjwood@ibm.com), [Jake Lishman](jake.lishman@ibm.com), [Joshua Skanes Norman](joshua.sn@ibm.com)    |
```
```suggestion
| **Authors**       | [Sam Ferracin](sam.ferracin@ibm.com), [Ian Hincks](ian.hincks@ibm.com), [Chris Wood](cjwood@ibm.com), [Jake Lishman](jake.lishman@ibm.com), [Joshua Skanes-Norman](joshua.sn@ibm.com)    |
```

SamFerracin committed:
Update 0022-circuit-block.md

Co-authored-by: joshuasn <53916441+joshuasn@users.noreply.github.com>

SamFerracin committed:
Update 0022-circuit-block.md

Co-authored-by: joshuasn <53916441+joshuasn@users.noreply.github.com>

SamFerracin committed:
Update 0022-circuit-block.md

Co-authored-by: joshuasn <53916441+joshuasn@users.noreply.github.com>

SamFerracin committed:
typos

SamFerracin committed:
Merge branch 'circuit-block' of https://github.com/SamFerracin/RFCs into circuit-block

ihincks commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,152 @@
+
+The concept of a block of gates is fundamental to many quantum computing routines. For example, a block is a natural concept when reasoning about quantum errors because the error profile often depends on factors such as which gates are applied simultaneously. However, Qiskit does not provide a dedicated object to represent a block. Traditionally, users have worked around this by defining blocks * ...
+
+To overcome these issues, we propose to establish the concept of a block of gates through the introduction of a new `Block` class. The primary goal of this class is to encapsulate an isolated block of `CircuitInstruction`s, treating it as a single unit for tasks like twirling and mitigation. We believe that `Block`s will offer:
```
```suggestion
To overcome these issues, we propose to establish the concept of a block of gates through the introduction of a new `Block` instruction class. The primary goal of this class is to encapsulate an isolated block of `CircuitInstruction`s, treating it as a single unit for tasks like twirling and mitigation. We believe that `Block`s will offer:
```

ihincks commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,152 @@
+  * Users will be able to draw blocks around every combination of instructions that they want to be grouped together for the purpose of twirling or mitigation (today, all blocks are effectively layers, i.e., depth-one blocks).
+* Ability to specify a twirling context:
+  * Users will be able to define a twirling group such as "Pauli" and a twirling strategy such as "active-accum" that can potentially be different for each block (today, this can only be done at the circuit level).
+* Ability to assign a noise handle to each block that uniquely specifies its noise model (today, this requires hashing circuits and is inherently unreliable, since circuits are mutable objects).
```
```suggestion
* Transparency: 
  * Users will be able to learn how their circuits are broken into blocks, twirled, and mitigated *before* running a job
  (today, they can only find this out once the job is done).
* Flexibility: 
  * Users will be able to draw blocks around any combination of instructions that they want to be grouped together for the purpose of twirling or mitigation (today, all blocks are effectively layers, i.e., depth-one blocks).
* Extensibility:
  * A block instance can own a context that can, for example, allow users to specify a twirling group such as "Pauli" and a twirling strategy such as "active-accum" that can potentially be different for each block (today, this can only be done at the circuit level). We can introduce allowed features to the context as necessary, and have definitions for what features the runtime supports.
* Ability to assign a handle to each unique block that enables user to target noise models against handles to be used in mitigation or simulation (today, there is no clean way of doing this, and usually involves an unsafe hash of a quantum circuit object, fragile indexing, or human introspection).
```

ihincks commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,152 @@
+1. In the existing implementation of blocks that exists in `pec-runtime`, blocks are hashable, so that they can be used as keys in dictionaries (specifically in the dictionary porduced by the `NoiseLearner` that maps blocks to noise models). This is clearly unsafe since circuits are mutable. How can we design the noise handle so that it is safer than that, but also more flexible (e.g., if users w ...
+2. Should `Block` be added to qiskit core, alongside the transpiler pass groups gates in blocks workflow 1? If so, should they be written in Rust or Python?
+3. What packages and subpackages need to be modified once `Block` is introduced? For example, should we add logic to qiskit-aer to be able to simulate circuits with blocks, or to Qiskit's optimizer, ..?
+4. How can we (should we?) facilitate re-use of CircuitInstruction or Block instances? For example, suppose we want the same block in a circuit 100 times:
```
```suggestion
4. How and when do we communicate the "scope of a block context"? For example, suppose a block context is only sensibly defined for depth-1 blocks of entanglers. Do we let this thing be instantiated with anything and raise during execution, or do we raise immediately? Who is in charge of validation?
5. How can we (should we?) facilitate re-use of CircuitInstruction or Block instances? For example, suppose we want the same block in a circuit 100 times:
```

ihincks commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,152 @@
+
+## `Block`: Definition and use cases
+
+A `Block` represents an isolated block of `CircuitInstruction`s that is treated as a single unit in the context of twirling and mitigation. In more detail, when a circuit contains blocks:
```
```suggestion
A `Block` represents an isolated block of `CircuitInstruction`s that is treated as a single unit in the context of twirling and mitigation. A circuit executor, such as the runtime primitives, should implement all like blocks acting on the same set of qubits in the same way. In more detail, when a circuit contains blocks:
```

ihincks commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,152 @@
+## `Block`: Definition and use cases
+
+A `Block` represents an isolated block of `CircuitInstruction`s that is treated as a single unit in the context of twirling and mitigation. In more detail, when a circuit contains blocks:
+- Twirling targets the blocks and ignores all the circuit instructions that live outside of the blocks.
```
However, the twirling interpreter responsible for execution will need to look at things outside the block and do things like single-qubit gate merging, no?

SamFerracin replied:
Yes, I agree with you

SamFerracin committed:
Update 0022-circuit-block.md

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

SamFerracin committed:
Update 0022-circuit-block.md

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

SamFerracin committed:
Update 0022-circuit-block.md

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

jakelishman reviewed (commented):
Thanks for this Sam.  I haven't read through it properly and fully marshalled my thoughts - I will try to do so tomorrow - but some follow-on questions that I think will shape more of what I'm thinking:

1. _After_ `pec-runtime`, how are we expecting `Block` to be handled as it's passed down to lower parts of a system stack?  "Flattened", or somehow passed down with existing structure?
2. What's preventing the scaling of `pec-runtime`'s current implementation of a `Block` as a custom `Instruction` subclass, and why might core Qiskit support be needed?  Is it more than syntactic convenience for the power-user case?
3. In the "power user" case, what guarantees about transpilation are being made by the user?  Are they required to build only ISA circuits if using `Block` manually, or are there more requirements in the transpiler?  Could you sketch out a complete "circuit construction to result retrieval" script that the user might write in both the power-user and non-power-user case?  I'm particularly interested in how/when the circuit becomes ISA safe, and how the interaction with a primitive will be.  I don't need any Qiskit-side or primitives-side implementation details, I'm just trying to get a handle on what you're envisioning the user side to look like, so I have a better idea of what's needed compiler-side.

ihincks commented:
1. The buck stops with whomever knows the next guy can't deal with it. Initially, all interpretation of twirling blocks would be done in the runtime: looking for blocks, constructing template circuits capable of executing the directives, sampling the parameters, passing the templates and parameters down. However, if it were known the next layer down knew how to interpret them (eg real-time twirling), then they would be passed, somehow.

2. We want to move all transpilation out of the primitives** so that it can focus on execution, and so that there is more transparency for users. This is impossible until users own blocks client-side. Since qiskit is our defacto programming paradigm, it seems like that ought to be the place where they are defined. If we wanted to have some of the context stuff reside in qiskit-ibm-runtime or be language agnostic, that would be on the table though. We also expect to have built in transpiler support for layering a circuit into blocks. I hope I'm getting at the question you asked.

3. I'm having a bit of trouble parsing `In the "power user" case, what guarantees about transpilation are being made by the user?` because the transpiler need not be touched at all in this case, but I'll plow through with an answer anyways.

```python
# POWER USER
from qiskit.blah.blocks.default_contexts import PAULI_TWIRLING

circuit = QuantumCircuit(127)
# the power user knows that block is in the target with support on any subset of qubits and PAULI_TWIRLING
with circuit.block(range(127), context=PAULI_TWIRLING) as layer:
    # the power user knows that cx(23, 27) is in the target
    layer.cx(23, 27)
circuit.measure_all()

sampler.run([circuit])

# CONVENIENCE USER
circuit = QuantumCircuit(2)
circuit.cx(0,1)
circuit.measure_all()

pm = generate_preset_pm(add_twirling_blocks=True, ...)
isa_circuit = pm.run(circuit)

sampler.run([isa_circuit])
```

** if something that needs to be done is most convenient to be done via the PM framework, then we reserve the right to do that. we just don't want it to be a requirement.

jakelishman commented:
Ok thanks Ian, that helps a lot.  I think you answered the questions I was _trying_ to ask, even if my wording wasn't super helpful.

Just to be clear: I'm totally on board with the general principles behind this proposal, I just have a bunch of considerations about exploring the design space, and making sure we're going for the right abstractions/APIs for the right time frames.

One more question, and then some responses:

* the way this proposal is written closes the door on the case of overlapping blocks (nested are allowed, but overlapping blocks at the same depth aren't).  Are we comfortable with that? ( Fwiw, I'm 100% comfortable with that, I just want to flag it as a design question.)

Responses:

1. ok, that's fine by me for now.  I'm particularly concerned with how this will interact with the future question of block re-use, though, and I don't want to introduce large API surface in the short term if there are less invasive alternatives, and the short-term form doesn't satisfy the long term needs.  I'll come back to this later.

2. That answer's roughly in line with what I was thinking, but let's say: ok, could we make do in the short term just by moving the custom `class Block(Instruction)` to be something that `qiskit-ibm-runtime` defines, rather than core Qiskit?  (Like point 1, I'll come back to this later.)

3. "the transpiler need not be touched at all in this case" is the answer I was looking for here, thanks.  That's the simplest form for us (and hence the script was easy to write).

Sorry, I don't have time to fully flesh out the "I'll come back to this later" bits today, but I'll try and get that done tomorrow.

---

Fwiw, with the current way control-flow context managers are done in Qiskit, there's no particular need to do `layer.cx` in the `with` statement; we'd likely just re-use the same logic for that, which would mean that you'd just use `qc.cx`, but it would get re-directed to the block builder.

jakelishman commented on 0022-circuit-block.md:
````diff
@@ -0,0 +1,153 @@
+            params=self.circuit.parameters,
+            label=label,
+        )
+```
````
[comment 2]

I think this should inherit from `ControlFlowOp`, not raw `Instruction`.  If it's raw `Instruction`, we have to re-invent a whole bunch of the special-case handling that was already super painful to teach the core library, except now it'll be slightly different as well, so even more complex.

As a few examples from the top of my head:

- all transpiler passes would immediately need to learn how to skip such an instruction
- ISA circuits checks would need special-case handling for this one instruction, since it doesn't behave the same way as other "container" objects that already exist
- the circuit visualisers would need to learn to handle this specially as well
- we'd have to invent new syntax for the block-construction phase in Qiskit, if we want to go beyond the "create another circuit and `append` it" form
- `assign_parameters` would need to be taught new logic to understand how to recurse into the block

jakelishman commented on 0022-circuit-block.md:
````diff
@@ -0,0 +1,153 @@
+## `Block` as an `Instruction`
+In this implementation, `Block` are `Instruction` objects that can be initialized from a circuit.
+```python
+class Block(Instruction):
````
[comment 1]

I would like to call this `Box`, since I think it aligns very closely with the OpenQASM 3 concept of a `box`.

jakelishman commented on 0022-circuit-block.md:
````diff
@@ -0,0 +1,153 @@
+        )
+
+        return self.append(Block(circuit, ...), qargs, cargs)
+```
````
This is a similar-but-not-identical version of what we already have for control-flow operations.

jakelishman commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,153 @@
+        )
+        cargs = tuple(
+            {b: None for carg in cargs for b in self._clbit_argument_conversion(carg)}
+        )
```
It's neither here nor there if we swap to using a `ControlFlowOp` (since the builders handle the arguments for you), but internally casting the inputs to a set-like object will mask problems or produce surprising results if the user submits `qargs` that accidentally name the same qubit more than once.  It's not super clear to me why we'd try and remove duplicates automatically here; why would the user _try_ to include duplicates in an ordered list?

jakelishman commented on 0022-circuit-block.md:
````diff
@@ -0,0 +1,153 @@
+```
+
+## Outstanding questions:
+1. In the existing implementation of blocks that exists in `pec-runtime`, blocks are hashable, so that they can be used as keys in dictionaries (specifically in the dictionary porduced by the `NoiseLearner` that maps blocks to noise models). This is clearly unsafe since circuits are mutable. How can we design the noise handle so that it is safer than that, but also more flexible (e.g., if users w ...
+2. Should `Block` be added to qiskit core, alongside the transpiler pass groups gates in blocks workflow 1? If so, should they be written in Rust or Python?
+3. What packages and subpackages need to be modified once `Block` is introduced? For example, should we add logic to qiskit-aer to be able to simulate circuits with blocks, or to Qiskit's optimizer, ..?
+4. How and when do we communicate the "scope of a block context"? For example, suppose a block context is only sensibly defined for depth-1 blocks of entanglers. Do we let this thing be instantiated with anything and raise during execution, or do we raise immediately? Who is in charge of validation?
+5. How can we (should we?) facilitate re-use of CircuitInstruction or Block instances? For example, suppose we want the same block in a circuit 100 times:
+```python
+for _ in range(100):
+    with circuit.block(context=my_context, name="foo") as block:
+        block.h(0) 
+        block.cx(0, 1)
+```
````
Some comments on the open questions:

1. making a block hashable within a context where you _expect_ to mutate them sounds like a recipe for horrible bugs at some point in the future.  If all you need is a handle to a block, and you don't want equivalent blocks to map to the same key, then just assign a UUID in the metadata of a block that you look up.  If you do, then define a "key function" that maps blocks to some UUID, and store the resulting UUIDs as the keys in the dict, rather than the blocks themselves.
2. imo, the `Box` I sketched above is suitable for Qiskit core, but not the pec-runtime-specific `Block` with `noise_key` and `twirling_strategy` that this RFC currently describes.
3. If we make this a `ControlFlowOp`, we'll be able to inherit a lot of the hardest work in other packages, but there's still some changes needed.  Off the top of my head:
   * Aer needs to be taught to effectively just unwrap the `block` as part of its control-flow-operation compilation.  This is easy.
   * `qiskit-ibm-runtime` and the IBM backends need to communicate that they support the control-flow op `box` in their `Target`s, just like they do for other control flow
4. These concerns sound very specific to pec-runtime's current implementation of twirling.  Certainly it seems inappropriate to constrain Qiskit core's implementation of `box` based on them.
5. As mentioned in the top comment, I really very much want to separate this idea of `box` from the idea of re-use.  There's really _loads_ of complexity in what it truly means to "re use" a structural element of a circuit (as opposed to just the unitary action of a gate, which is the sort of re-use we handle nowadays).  I think this `box` should be considered entirely separately to the question of re-use, and we should make it clear that `box` is the primitive that solves "grouping with a context" and not the primitive that solves the (somewhat unrelated) problem of "reuse of program structure".

ihincks replied:
4. Issue 4 can be more generally stated as: "can the annotations of a box have an impact on the allowed contents of a box at construction time?" Offline, a subset of us discussed whether or not this kind of thing would be more appropriate during primitive validation.

5. I'm not sure exactly what you mean by reuse. The Box concept being discussed is reusable in the sense that, like every other kind of instruction, there would be nothing stopping someone from adding the same box instance multiple times to a single circuit. By reuse, do you mean that tools such as the transpiler would be forced to perform the exact same operation on every like block encountered? Like if it scheduled the contents of a box in one way, if it ever encountered the same block again, it would by contract have to schedule the new one in the same way?

jakelishman reviewed (commented):
Coming back from that previous comment, and with me having thought about it a lot more:

## Rough concerns

My initial worries about this RFC are that I agree with the summary statement of the RFC about blocks being natural and necessary, but that I don't think this RFC gets enough into the semantics of precisely what a block means in a circuit IR, and so how the transpiler infrastructure should handle it.  This RFC currently largely punts on those questions (the advanced user is required to submit an ISA circuit without transpilation, etc), but if the type is coming to core Qiskit, I think it's necessary for us to worry about that now.

I'm very concerned that we don't introduce new syntax, objects, or special cases into the transpiler with the intention of making an object general later, only to discover that we can't make it more general (the open questions), so I'm keen to design this new object with the idea that it'll be limited, not sort of open to the future, and ideally to re-use as much as already exists in Qiskit as possible.  Everything new added to the library (especially structural things) has a big risk: they increase the complexity that users have to understand, and the logic that consumers have to interpret; they risk the API feeling disjointed and ill-fitting; they risk tying our hands for new features in the future.  Of course that's then weighed by the utility of what it is we're adding.

Just as some background to my thinking too: my initial reaction to this was that I didn't want to add a special-case block structure to Qiskit with the semantics listed in this RFC, because I conflated the "blocks" here very heavily with the idea of reusable blocks, and imo we have very large unanswered questions about how those will semantically work with transpilation (how will we route a 3+q block, how will we handle heterogeneous ISAs within them, how will we handle scheduling variations between different 2q links, etc etc).  I want to make sure that the block we introduce here doesn't get in the way of proper re-usable blocks later.


## Design thoughts

Ok, so that said, I totally agree with the summary statement in the RFC, and I think a specifically _non-reusable_ block still has a lot of merit, distinct from the idea of a subroutine-call-like reusable block.

With my phrasing above, I want to identify this with OpenQASM 3's `box` object, that is:

- a non-overlapping (but potentially nested) set of circuit instructions
- that introduces a new scope for declared variables,
- which can have internal scheduling within itself, but is scheduled atomically within the circuit (i.e. all wires touched within the box are delayed to match the critical path),
- which forbids optimisations from crossing the boundaries of the box,
- and can have arbitrary backend- and SDK-agnostic annotations attached to it.

Syntactically, within Qiskit, I want to re-use all of the existing control-flow mechanisms to do this.  Section "`Block` as an `Instruction`" of this RFC introduces new syntax that's similar-but-not-the-same as the existing control-flow structures, and a `box` can be interpreted as the most trivial of control-flow structures - in a CFG, it's got a single in edge and a single out edge.

This is fairly minor, but I'd like to call the object "box", not "block", because block is way overloaded already in terms of CFGs (and Qiskit already has the `ControlFlowOp.blocks` attribute), plus OpenQASM 3 already defined the term `box`, and I'd like to promote that association so it's easier for users.

The main differences from the RFC that this would imply would be:

- [comment 1] we call it `Box`, not `Block`
- [comment 2] we inherit from `ControlFlowOp` rather than raw `Instruction`

I'm not at all keen to raise `twirling_strategy` or `noise_handle` to the level of Qiskit core here.  I do believe they're important, but I think that's a level of API coupling that'll totally get in the way of using the object, and extending APIs in the future.  What if somebody _doesn't_ want to twirl, but just wants to do ZNE with the `box` as an atomic instruction unit?  What if they do want to twirl, and pec-runtime wants to introduce a new experimental twirling strategy in a shorter time-frame than Qiskit's 3-month release window?  (and so on).

Instead, I want to make this much more general, and include a system in this RFC for how a user, compiler, and backend-specific plugin can create and work with "contexts", how they should be handled by ISA-circuit verification, how we'll serialise / deserialise them through QPY and how we might import / export them through OpenQASM 3.

I don't have a full design for this in mind, but here's a rough sketch of what I'd envision construction looking like:

```python
from qiskit.circuit import QuantumCircuit

# Hypothetical core-Qiskit box-level annotation.
from qiskit.transpiler.contexts import ScheduleStrategy

# Hypothetical annotations for IBM-specific primitives.
from qiskit_ibm_runtime.contexts import TwirlingStrategy, NoiseModel

qc = QuantumCircuit(5, 5)
with qc.box([ScheduleStrategy.ALAP, TwirlingStrategy(model=NoiseModel(...))]):
    qc.h(0)
    qc.cx(0, 1)
    qc.x(2)
with qc.box([ScheduleStrategy.ASAP, TwirlingStrategy(...)]):
    qc.h(2)
    qc.cx(2, 3)
    qc.x(4)
qc.measure(qc.qubits, qc.clbits)
```

`QuantumCircuit.box` would be the same kind of hybrid control-flow builder that `QuantumCircuit.if_test` and friends are.  The created instruction would be something like

```python
import typing
from qiskit.circuit.controlflow import ControlFlowOp

@typing.Final
class Box(ControlFlowOp):
    def __init__(self, circuit, annotations):
        super().__init__("box", circuit.num_qubits, circuit.num_clbits, [circuit])
        self.annotations = annotations

    @property
    def blocks(self):
        return (self.params[0],)

    def replace_blocks(self, blocks):
        assert len(blocks) == 1
        return Box(blocks[0], self.annotations.copy())
```

### Semantics of raw `box`

Now we have to define the exact semantics of `Box` to various parts of Qiskit.  I think there's a couple of points about `Box` that I can say fairly non-contentiously (not accounting for annotations, so far):

- `box` is a structural element that allows shadowing of `Var` nodes within it.
- a `box` is _in some ways_ atomic, in that an optimisation pass cannot cancel a gate inside a `box` with a gate outside, but may choose to commute a gate all the way through a box.  We have `barrier` already if the user additionally wants to forbid commutation all the way through (they'd put a barrier outside the box on one side).
- by default, we should expect that the compiler will enter `box`es, and treat them as it would any component of a circuit - it should basis-translate instructions, make sure they're routed, and run its standard optimisation loop within the box.

Next, I think there's a question where it gets a bit more contentious (still, without talking about annotations):

- OpenQASM 3's `box` is atomic from the perspective of scheduling as well.  That is, the box has its own duration (equal to the critical path of the box), and all qubits within it are delay-padded within it to satisfy that (things like DD can split delays within the box, they can't apply across the box boundary).  Qiskit core so far has not really considered scheduling to be as important a compiler function as others, mostly for organisational reasons - that component was handled by other teams for the IBM stack, and the `Target` model doesn't give enough information to do this entirely properly yet.  That said, I think making `box` atomic with regards to scheduling as well is probably the right call: the OQ3 interop story would be nasty without it; it's easier to explain and reason about atomicity if it applies to scheduling too; it's probably what's expected anyway.  The risks are mostly that overuse of `box` can reduce the effectiveness of DD-like passes.

So those bits above are about the design of "raw" `box`.  For inclusion in Qiskit core, these are really important for us to start with.  I think the points, as I've laid them out above, are entirely compatible with pec-runtime's current requirements.


### Arbitrary annotations

I've used "annotations" in everything I've written so far because that's what OpenQASM 3 calls this concept.  What I mean is a "compiler directive scoped to this `box`", and you can have multiple annotations, which are explicitly order-independent, on the same `box`.

So now this is where it gets hairier.  I think it's best for everybody if pec-runtime's error-mitigation semantics are defined and owned by the IBM primitives team, not by core Qiskit (I've given enough reasons elsewhere, but can expand if desired).  The problem with that, though, is that now we have to define an abstraction layer for what an annotation can mean, and how a compiler pass is permitted to behave in the presence of an annotation it doesn't understand because it's defined in some unknown package.  From a practical perspective for IBM's implementation, we also need to consider how such annotations will be serialised/deserialised over the network boundary without allowing/requiring arbitrary code execution (e.g., how do we make this work with QPY?).

#### Semantics of all annotations

I can imagine that we (in the future, not in this RFC) supply some core directives, such as the hypothetical `ScheduleStrategy` one I had above that affect Qiskit's built-in compiler passes, but more generally, I want to think about these as defined by _any_ compiler pass author, and I want to set the API so that other passes know what they may/may not do with an annotated `box`.  The extremes are:

1. any compiler pass is allowed to entirely ignore any annotation it doesn't understand, provided it leaves the annotation in place for a later pass.
2. a compiler pass is forbidden to act on a `box` unless it declares that it understands all annotations attached to it.
3. the _annotation_ somehow declares what compiler passes are allowed to run on the box.

Point 2 would be fairly unsatisfactory for core Qiskit - our core transpiler passes can't be aware of all the possible annotations out there, and so we'd end up limiting custom-annotation authors to either do all the compilation work Qiskit usually does for them, or require their users to do that instead.  That's obviously no good from an API perspective.

Point 1 is limiting because it means a backend-specific author can't have an annotation that lets the `box` use very backend-hyper-specific stuff without risking another compiler pass mess it up.  For example, let's think to the future when backends are all error-corrected and defined in terms of logical qubits, but a backend wants to use some sort of "verbatim" box to allow temporary and very low-level temporary access to the physical qubits (something akin to inline `__asm__` blocks in gcc/clang-flavoured C).  The core of the idea is that a compiler directive might be either an additive "apply this additional behaviour" marker, _or_ it might be a subtractive "require these temporary restrictions" marker; taking point 1 or point 2 as the default is akin to choose either the former or the latter as the only type of directive permitted, respectively.

Point 3 alone is probably not great for the same reasons as point 2; an annotation can't know all possible passes.  That said, I think a limited subset of this might be possible: an annotation might mark itself as "allow me to be ignored" or "require me to be understood". I'd want to think more about this before committing to it, but I think it's plausible (really, I think I want more infrastructure around semi-general circuit "predicates" all the way through the compiler, but that's a much larger question than can be answered here).

We do also have a potential option in that core Qiskit can specify a small number of "special case" annotations, which _all_ passes are required to interpret and respect.  This gets tricky from a backwards compatibility perspective, though - if we add a new one of these special annotations in Qiskit 1.3, say, it might invalidate code that worked happily in Qiskit 1.2.


#### Serialisation and deserialisation

If we allow arbitrary annotations, we reasonably need some way of constructing / deconstructing these, without sending arbitrary code to be executed over the wire - that'd be a huge security hole.  This is the problem that QPY solves for circuits.  The general principle is that we can rely on reconstruction on the far side of a network boundary, if the receiver only runs code that it chose to install ahead of time. `pickle` is not suitable for this.

In my model here, I think we'd maybe need some sort of plugin interface from QPY, so packages that provide a compiler annotation register with QPY the way to serialise / deserialise it.  Both sides of the QPY stack would need to have the same plugins installed to be able to import the resulting circuit, and that means that when the deserialiser is deployed in some remote container, all the deserialiser code can happily be audited as normal to verify it's safe.

The risks I see here are that introducing a plugin system fits very awkwardly with QPY's _very_ strict backwards-compatibility guarantees.  If we are to maintain them, we'd have to provide, at a minimum, a bunch of guidance for how annotation authors should write their serialisers/deserialisers to maintain those guarantees.  Maybe there's a better way.

The same sort of logic can be used for import/export from OpenQASM 3, but since that's lossy and we certainly can never make the same guarantees about backwards compatibility with that, it's rather less of a concern to me.

## Final thoughts

I've sketched out how I'd generalise this RFC to make it suitable for inclusion in core Qiskit.  Notably, I've not attempted to define any sort of semantics for `twirling_method` or `noise_handle`, but only attempted to put enough pieces in place that the IBM primitives team (or anybody building on top of Qiskit) can do that themselves.

I want to emphasise that it's a _sketch_: I think it's generally a good, scalable idea - and overall I think actually a lot less work to implement than the current RFC because it re-uses a lot more of the existing Qiskit infrastructure - but I've not polished it, and I think others should try and poke holes in it.

ihincks commented:
Hi @jakelishman thanks for your detailed comments that are, as always, insightful. 

We have no problems with:
>
    [comment 1] we call it Box, not Block
    [comment 2] we inherit from ControlFlowOp rather than raw Instruction

We also have no problems taking ownership of twirling stuff inside of `qiskit-ibm-runtime` as special annotations. It seems conceivable to me that having some kind of UID annotation be a part of qiskit-core might still be on the table. If you recall my 2023 presentation, I framed twirling directives in terms of qasm3 annotated boxes, so this is not inconsistent with what my thoughts have been for a while.

It sounds like we want to reword the entire RFC as the introduction of `Box`, with a main goal of making sure that it is compatible with using it for twirling infrastructure, and whatever else you have in mind. Any mention of twirling should be clearly labelled as an example usage, and not confused with the promise of the RFC.

I know you were just throwing around examples and not prescriptions ("I want to emphasise that it's a sketch"), so I bring it up only in the context of serialization: I would not want to support something like `TwirlingStrategy(model=NoiseModel(...))`, I really dislike having noise-like things living inside of circuits. Instead I'd want the ability to label boxes by u/uids and have the noise model attach against those, separately. So having annotations required to, for example, have a json-able format, would not feel limiting to me presently.

Interactions with the transpiler (forgive me, but I never call it a, and especially not "the", compiler to reduce naming confusion between teams) need further thought, and your section about it is a great start.

jakelishman committed:
Fix metadata

jakelishman committed:
Pivot RFC to generic framework for `box` and annotations

This fleshes out a lot more details about how a generic `box` would work
in Qiskit SDK, and how arbritrary downstream annotations might work in
the context of an entire compiler framework, where annotation-creating
passes may be injected inbetween standard Qiskit passes, or the user may
write into their circuit, and the transpiler might want to verify before
submission to the quantum computer.

The case study on how `pec-runtime` will use this isn't fleshed out - I
mostly just moved it into a section at the end and left it.

SamFerracin committed:
Update 0022-circuit-block.md

SamFerracin committed:
Update 0022-circuit-block.md

SamFerracin committed:
Update 0022-circuit-block.md

SamFerracin committed:
Update 0022-circuit-block.md

SamFerracin committed:
Update 0022-circuit-block.md

SamFerracin committed:
Merge pull request #2 from jakelishman/circuit-block-to-box

Pivot RFC to generic framework for `box` and annotations

SamFerracin commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,300 @@
+We also do not currently have the concept of a no-op on classical data.
+To make the user intent clearer here, we propose adding a `QuantumCircuit.noop` to make the intent explicit:
+
+> [!NOTE]
+> *Bikeshedding potential*: do you prefer `QuantumCircuit.use` or `QuantumCircuit.noop` for this?
+> `use` is perhaps clearer for more users, but its name only makes sense in a control-flow builder context, whereas `noop` makes sense to appear in the root scope of a `QuantumCircuit`.
+
```
We should probably remove this before opening the PR

```suggestion
```

ihincks replied:
i vote to open the rfc with `noop`

SamFerracin committed:
block --> box

SamFerracin committed:
better title

SamFerracin commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,300 @@
+    - They submit their job.
+
+Setting up these two workflows presents notable advantages, such as:
+- All of the existing transpiler passes can be asked (at least initially) to simply ignore the boxes, since transpilation is meant to happen before the boxes are generated. This can save quite a lot of time that would otherwise be spent adding logic to every single transpiler pass -- and yet, we can choose to do this in the future.
```
This can probably be removed, now that we use control flow ops?

joshuasn replied:
it might be worthwhile to instead say `to simply ignore the boxes with twirling annotations` or some verbiage to connect it back to [this line](https://github.com/SamFerracin/RFCs/pull/1/files#diff-8468c13b6726725a9dfdf5968d828e7a52b13e30d77470e9e6f50d126179d56aR254)

jakelishman replied:
I didn't touch the text in this section - when I'm back, I can review any changes in here that you want to make in the meantime?

joshuasn commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,300 @@
+
+Setting up these two workflows presents notable advantages, such as:
+- All of the existing transpiler passes can be asked (at least initially) to simply ignore the boxes, since transpilation is meant to happen before the boxes are generated. This can save quite a lot of time that would otherwise be spent adding logic to every single transpiler pass -- and yet, we can choose to do this in the future.
+- Users of the primitives can inspect the boxes before submitting their jobs. This is a big improvement: today, the box generation happens in the server and can be unintuitive.
```
again, maybe reword in terms of annotations

ewinston commented on 0022-circuit-block.md:
````diff
@@ -0,0 +1,300 @@
+```
+
+For `box` in particular (but also for other control-flow operations), it can be useful to mark certain qubits and other circuit resources as being explicitly "used" by a block, even if the actual operation is a no-op.
+For example, twirling and noise learning may want to consider certain qubits along with a complete block, even if that qubit is undergoing the logical identity for the span.
````
If a box is control-flow block of some width with a non-zero number of idle qubits the transpiler doesn't reduce the width of the circuit does it? It may cause idle qubits to become non-idle when transpiling within the block, however, so is the noop an indication to preserve idles for the duration of the box?

joshuasn replied:
No, it won't reduce the width of the circuit. Passes will leave the box untouched until a pass that understands the annotation encounters it. The noop is a signal that the qubit it acts on can be touched in the box.

ewinston reviewed (commented):
One of the things that's not clear to me is whether the box needs to exist as a separate structure on the QuantumCircuit as is currently done with ControlFlowOp. In particular the requirement of not optimizing across box boundaries seems like it could complicate fair optimizations. For instance:
- 1q run twirling
- DD sequence application semantics would likely have to change or need special casing e.g. when an idle period crosses a box boundary. In this case the DD pass would attempt to put a DD sequence before the box and at the beginning of the box, so what may have initially been a single DD sequence would become two DD sequences. Depending on value of `sequence_min_length_ratio` it's possible one or both of the split delays could even be too short to apply a DD sequence resulting in uncorrected errors.  

It seems like a box could instead exist as an annotated collection of instructions and qubits without existing as a QuantumCircuit instruction itself. In a simple view it could be described by something like a set of references to instructions in QuantumCircuit/DAGCircuit. This would seem to satisfy many of the requirements for boxes described:

- describes a collection of instructions for twirling
- can be ignored by transpiler passes which don't understand them: since they would exists outside QuantumCircuit's instructions, transpiler passes wouldn't even know they exist unless they checked the location where transpiler annotations are stored (e.g. not QuantumCircuit.data). 
- would be simple to serialize since they are just some lightweight reference to existing instructions.

Additionally such a structure could also possibly be used to annotate other high-level properties of the QuantumCircuit which could be useful for other transpiler optimizations and which may overlap in the future when rules for dealing with these annotations are elaborated on. 

As far as an interface goes it can be implemented as a context manager as already suggested and so the examples of usage described in this RFC would not have to change.

SamFerracin commented:
Thanks for your reply, @ewinston 😄 
A few remarks about annotated collection VS boxes:

1. I don't fully understand why "the requirement of not optimizing across box boundaries seems like it could complicate fair optimizations [for twirling and DD]," and it's unclear how the collection would get around this.

    a. At the end of the day, boxes is what we do in the server (we turn the circuit into alternating sequences of one-qubit gate layers and two-qubit gate layers, and then start appending twirling gates and all that). Boxes expose this process, so that it is done transparently and the user can understand (or even define) how the layers are generated before submitting the circuit. In other words, we do not lose anything as compared to existing functionality, but we gain transparency and provide even more functionality than is available today.

    b. The box provides a clear interface where whenever we use the transpiler, if the annotation of the box is of a certain type that matches the pass(es), the transpiler is allowed to get inside the box and do all the modifications. It's unclear how a collection would be dealing with transpilation any differently

2. The annotated collections seem (to me, at least) a more complex and less intuitive option:

    a. More complex: Having instructions AND references to instructions means that we constantly need to:
        - check if a dag node is or is not part of the collection
        - update both of them.
     For example, imagine replacing a node in the dag circuit with another node. We need to replace the node, then check if it was an annotated one, and if it was, replace it in the collection. Two sets of operations instead of one (with the box)

    b. Less intuitive: As a user, I can see a box just by drawing the circuit, as opposed to having to look at the collection. I also think that it will be easier to document ("this is a box, this is what happens"), as opposed to document what's happening with an annotated collection that lives on top of the circuit and that a user doesn't really interact with

ewinston commented:
An annotated collection could have an issue with having to check the annotation sets whether it is a member (O(1)) and whether node editing is allowed (O(1)) which box does not seem to have to worry about. 

Perhaps I can pose a couple of situations which I'm wondering how the box instruction would handle. Let's say the user defines the following circuit where the box of the second cx gate has been draw to include neighboring single qubit gates which can be folded into the twirls. 
![box_pre_opt](https://github.com/user-attachments/assets/4078ca42-f730-4e88-9d95-0ff4afdca287)

After passing through qiskit's transpiler additional single qubits gates (or even swap gates if the circuit isn't routed) may be added which may change how the user would draw the box.
![box_post_opt](https://github.com/user-attachments/assets/7781be5c-a61c-488f-a1dd-c5009558caff)

How can circuits which introduce single or two qubit gates after the box is defined be optimized?

The second scenario can be imagined from the above image. Suppose the user applies DD after defining the box. On qubit q2 above the dd sequence might use the time up until the box boundary whereas in the absence of the box boundary it may have used up to the cx within the box. Is there a way to not effect the behavior of the DD sequence across the box boundary?

SamFerracin commented:
@ewinston 

A box is a directive that allows users to say "leave these gates alone." So in your example, where the user is intentionally placing the H gate inside the box, the transpiler is providing the best optimization that it can provide, given the constraint placed by the user. The same applies to DD.

If anything, this was a poor choice on the user's side (they should have left the H gate outside of it to take full advantage of transpilation and noise suppression), but it is their choice. The consequences of this choice are clear and leave no room for interpretation.

I also want to point out that the "basic" workflow (the user utilizes a compiler pass that adds the boxes for them) will not set boxes up like this, it will leave one-qubit gates out (I mean.. It's up to us to write that pass, we'll write the most optimal one). So the circuit that you are considering is coming from the "advanced" workflow, where the user creates the boxes on their own and deliberately makes that choice for H

ewinston committed:
add section about dynamical decoupling with boxes

ihincks commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,300 @@
+
+* a block of several gates might have a known unitary action, and can be treated as atomic for commutation purposes, even if a candidate gate doesn't commute with each block element individually.
+* noise-learning techniques might be applied to simultaneous logical blocks of operations, rather than individual 2q hardware gates.
+* a user might wish to Pauli twirl around a block of instructions, rather than a single one.
```
```suggestion
* a user might wish to Pauli twirl around a block of instructions, including idling wires, rather than a single one.
```

ihincks commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,300 @@
+* a block of instructions might be assigned some duration as part of a larger circuit, and internally it is scheduled with stretchable durations to implement optimal dynamic decoupling spacing without increasing the complexity of the outer scheduling problem.
+
+Several downstream projects of Qiskit have made ad-hoc blocks by using labelled `Barrier` instructions, but this optimisation barrier has further effects than simple grouping, and the single-sided nature of it makes it very hard to work with as an actual "grouping" construct.
+`Barrier`s also are very hard for users to interpret when inspecting circuits after compilation.
```
superfluous superlative
```suggestion
`Barrier`s also are hard for users to interpret when inspecting circuits after compilation.
```

ihincks commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,300 @@
+Several downstream projects of Qiskit have made ad-hoc blocks by using labelled `Barrier` instructions, but this optimisation barrier has further effects than simple grouping, and the single-sided nature of it makes it very hard to work with as an actual "grouping" construct.
+`Barrier`s also are very hard for users to interpret when inspecting circuits after compilation.
+
+The RFC proposes to add a new `Box` instruction to Qiskit, whose semantics will be very similar to the `box` concept from OpenQASM 3.
```
```suggestion
The RFC proposes to add a new `Box` instruction to Qiskit, whose semantics will be very similar to the `box`
[concept from OpenQASM 3](https://openqasm.com/language/scope.html#block-scope).
```

ihincks commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,300 @@
+
+As in OpenQASM 3, a `Box` will be:
+
+* a group of non-overlapping (but potentially nested) circuit instructions,
```
I'm not sure that the OQ3 spec actually does talk about contained instructions being non-overlapping, and I'm not sure why it would for the same reason I wouldn't expect an `if` block to have such limitations. I expect this kind of thing to fail during a validation phase when, for example, a box annotated with pauli-twirling attempts to have overlapping entanglers.

```suggestion
* a group of possibly-nested circuit instructions,
```

jakelishman replied:
Yeah, the sentence as I wrote it is a mess.  I moved the bit about nesting out of the list, and modified the top bit to call them "consecutive and/or parallel instructions", though there's probably some more formal graph-theory terminology that I can't bring to mind.  Done in 5e579ba.

ihincks commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,300 @@
+* which forbids optimisations from crossing from outside to inside the box or vice-versa, but permits optimisations across the entire box,
+* and can have arbitrary backend- and SDK-agnostic annotations attached to it.
+
+A `box` can contain any other Qiskit circuit instruction.
```
Redundant
```suggestion
```

ihincks commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,300 @@
+* and can have arbitrary backend- and SDK-agnostic annotations attached to it.
+
+A `box` can contain any other Qiskit circuit instruction.
+Aside from variable scoping concerns and the resolution of delay lengths, the execution of a circuit containing a `box` should be the same as the execution of the same circuit with the boxed instructions inlined into the circuit.
```
This sentence doesn't look True. for example,

```
circuit.h(0)
with circuit.box():
    circuit.h(0)
    circuit.h(1)
```

has no variable scoping or delays, the only two caveats mentioned, but I might reasonably expect it to be equivalent to 

```
circuit.h(0)
circuit.barrier(0, 1)
circuit.h(0)
circuit.h(1)
```

rather than

```
circuit.h(0)
circuit.h(0)
circuit.h(1)
```


at execution time.

So, does the sentence mean to put "logical" in there somewhere, or should it talk a bit more about barrier sandwiches?

jakelishman replied:
Fixed in 5e579ba to describe the walls as synchronisation points with no associated duration.

ihincks commented on 0022-circuit-block.md:
````diff
@@ -0,0 +1,300 @@
+    qc.noop(*qc.qubits, *qc.clbits)
+```
+
+The `noop` (or `use`) instruction will do the standard Qiskit circuit error checking to validate its inputs are valid resources for the circuit, but will not cause a new entry in the instruction list.
````
```suggestion
The `noop` instruction will do the standard Qiskit circuit error checking to validate its inputs are valid resources for the circuit, but will not cause a new entry in the instruction list.
```

ihincks commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,300 @@
+Two ways of making this extension would be:
+
+* have a plugin interface to QPY, where installed packages can register additional serialisers / deserialisers for annotations.
+* add additional keyword arguments to the `dump`/`load` functions that allow a user to pass in the additional serialisers / deserialisers manually (like how is done for JSON).
```
i'd vote for the kwargs on dump/load for custom serializers.

since the same annotations would be showing up repeatedly in the same circuit, perhaps it could cache by id like pickle.

jakelishman replied:
I added a note mentioning your (and my) preference, but explicitly punted the final decision out of this RFC.  We have the main outline of how to do this, and we can flesh out the details fully during implementation on Qiskit.

ihincks commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,300 @@
+
+> [!NOTE]
+> This is a _very vague_ sketch, I'm just trying to make sure we have a clear definition of what a `Target` is, and we don't start having it become so abstract that it is no longer a good model for a QPU.
+> This is absolutely not necessary to finalise for this RFC, but I feel like it's worth thinking about this as a joined-up story, since previously we have treated the output of `transpile` as "ready for QPU execution", but annotations stymie that.
```
There are other things that stymie that too at the moment, like parameter expressions, which the quantum computer handles but the QPU doesn't. All in all, I really like the direction this section is going in terms of making the distinction between QPU<->Target, Quantum Computer<->Backend nice and crisp.

As for the `sample` and `estimate` methods on backend, I think my biggest concern here is whiplash: we just yoinked `run` away, and this would look like backpedaling. It seems like we could alternately make primitives V3 have to own a backend V3 instance.

jakelishman replied:
(For tracking purposes: this is effectively a duplicate thread of https://github.com/SamFerracin/RFCs/pull/1#discussion_r1876733184.)

ihincks commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,300 @@
+
+## Case study: twirling, noise learning, and mitigation with `qiskit-ibm-runtime`
+
+We envision two main workflows that users will follow once `Box`es become available, one for regular users and another one for power users:
```
I don't think this sentence should be talking about `box` users in generality, it should be specifically talking about twirling box users; generic box users' workflow might look totally different.
Instead, this case study needs a short paragraph intro that is twirling/learning/mitigation specific. eg "An application for annotated boxes is to declare to an execution stack that a set of operations should ..blah blah blah"

ihincks replied:
Reading through the whole section, I think we need to:
 - add examples (with disclaimer that we are not attempting to define twirling annotations in this PR, but just give a flavour)
 - better connect to everything written above, especially transpiler passes. what happens with a layout pass if a user *does* happen to create a twirling box manually and then invoke the stock pm?

ihincks commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,300 @@
+- All of the existing transpiler passes can be asked (at least initially) to simply ignore the boxes, since transpilation is meant to happen before the boxes are generated. This can save quite a lot of time that would otherwise be spent adding logic to every single transpiler pass -- and yet, we can choose to do this in the future.
+- Users of the primitives can inspect the boxes before submitting their jobs. This is a big improvement: today, the box generation happens in the server and can be unintuitive.
+
+In addition to supporting the two workflow above, we believe that the following workflow should *not* be supported:
+
+3. The workflow for "lazy" users:
+    - Users initialize a `QuantumCircuit` without boxes, as they do today.
+    - They apply all of the desired transiler passes, for example to map the circuit to an ISA circuit for the backend that they wish to use.
+    - They submit their job.
+    - The server creates boxes for the users "behind the scenes."
```
while important, this comment i think has now strayed far from the goal of this RFC since it's about the runtime service implementation and features rather than boxes and annotations. we could instead highlight that boxes bring the possibility of making such operations transparent while still allowing the tight sampling loops to happen near-time

SamFerracin committed:
Update 0022-circuit-block.md

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

SamFerracin committed:
Update 0022-circuit-block.md

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

SamFerracin committed:
Update 0022-circuit-block.md

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

SamFerracin committed:
Update 0022-circuit-block.md

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

SamFerracin committed:
Update 0022-circuit-block.md

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

ihincks committed:
Changes to twirling/mitigation section

> Co-authored-by: Samuele Ferracin <sam.ferracin@ibm.com>

SamFerracin committed:
Merge pull request #4 from ihincks/edits

Edits

joshuasn commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,337 @@
+Twirling is the process of taking a single base circuit and randomly inserting gates from some distribution, composing them with existing gates, such that the unitary implemented by the circuit and its structure are unchanged. When data is averaged over the result of executing many such independent randomizations, this has the effect of modifiying the noise profile of the circuit into a more trac ...
+
+There are many choices for how one might choose to perform these randomizations, where the random gates should be inserted, and with what existing gates they should be composed with.
+If this process is completely automated by the execution stack, the the user needs to be able to clearly specify how these choices should be made.
```
```suggestion
If this process is completely automated by the execution stack, then the user needs to be able to clearly specify how these choices should be made.
```

joshuasn commented on 0022-circuit-block.md:
```diff
@@ -272,21 +272,58 @@ class BackendV3:
 
 
-## Case study: twirling, noise learning, and mitigation with `qiskit-ibm-runtime`
+## Case study: twirling, noise learning, and mitigation
```
Maybe out of scope but could also talk about how PEA can be represented in this way? That is, emitting a Pauli error with rates multiplied by `num_amplifications`?

I remember there was some work @SamFerracin did where only a subset of entangling gates were amplified, this could be a way of specifying said subset (though deterministically as opposed to randomized here).

SamFerracin committed:
Update 0022-circuit-block.md

Co-authored-by: joshuasn <53916441+joshuasn@users.noreply.github.com>

ihincks commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,337 @@
+Therefore, one application of boxes and annotations is to let users declare how they would like to perform twirling, for example with respect to how the gates are randomly drawn, where they are inserted, and what they are composed with.
+Once a circuit has been decorated with these boxes and annotations, it is a self-contained description of how twirling should take place.
+
+> **_NOTE:_**  In what follows, we attempt to give a flavor of what twirling annotations might look like in `qiskit_ibm_runtime`. They will not be defined by the `qiskit` library itself. Take the particular syntax or conventions of the annotations as preliminary.
```
```suggestion
> **_NOTE:_**  In what follows, we attempt to give a flavor of what twirling annotations might look like in `qiskit_ibm_runtime`. They will not be defined by the `qiskit` library itself. The particular syntax and conventions used here are preliminary and do necessarily represent what will come.
```

ihincks commented on 0022-circuit-block.md:
````diff
@@ -0,0 +1,337 @@
+    circuit.measure_all()
+```
+
+We envision two main workflows that users will follow once `box`es and annotations become available to specify twirling, one for regular users and another one for power users:
````
```suggestion
We envision two main workflows that users will follow once `box`es and annotations become available to specify twirling, one for regular users and another for power users:
```

SamFerracin committed:
uuid

SamFerracin committed:
Merge branch 'circuit-block' of https://github.com/SamFerracin/RFCs into circuit-block

SamFerracin commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,364 @@
+  
+This workflow essentially takes the same steps as are taken today and therefore presents the same disadvantages discussed above, and also violates the principle "don't do hidden transpilation server-side".
+
+Another application of boxes is to let users declare how they would like to perform noise learning and mitigation. To do so, they can assign a unique identifier to a box, and subsequently use those identifiers in the primitives' options.
```
@ihincks FYI

ihincks replied:
we can probably be slightly more specific

> Another application of boxes is to let users declare how they would like to perform noise learning and mitigation. To do so, they can assign a unique identifier to a box via a separate UID annotation, and subsequently provide the primitives' options with a map from those identifiers to noise models they wish to attach to the identifier.

SamFerracin committed:
Update 0022-circuit-block.md

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

SamFerracin committed:
Update 0022-circuit-block.md

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

SamFerracin committed:
CR

SamFerracin committed:
Merge branch 'circuit-block' of https://github.com/SamFerracin/RFCs into circuit-block

ihincks commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,359 @@
+    def target(self) -> Target: ...
+    @abstractmethod
+    def allowed_annotations(self) -> AnnotationChecker: ...
+    @abstractmethod
+    def sample(self, sampler_pub_likes) -> SamplerJob: ...
+    @abstractmethod
+    def estimate(self, estimator_pub_likes) -> EstimatorJob: ...
```
My biggest concern here is perceived whiplash from users who were migrated away from `backend.run` to separate primitive classes, only to be ushered back to backend execution methods. 

I do agree, however, that having the correspondence QuantumComputer feature set <-> Backend, QPU feature set<-> Target is nice and clean.

jakelishman replied:
I need to remove this bit that mentions mixing-in of the "execution" methods (`sample`, `estimate`) - it was part of what helped me structure things in my mind on the first pass, but leaving the text there is just going to be a huge distraction from the RFC.  As you say, we can't actually put the execution methods back onto the `Backend` class unless we wanted to alienate every user who just migrated.

jakelishman replied:
Done in f2a6e5c.

yaelbh commented on 0022-circuit-block.md:
```diff
@@ -0,0 +1,359 @@
+* a group of non-overlapping (but potentially nested) circuit instructions,
+* which introduces a new scope for declared variables,
+* which can have internal scheduling within itself, but is scheduled atomically within the circuit (i.e. all wires touched within the box are delayed to match the critical path length),
+* which forbids optimisations from crossing from outside to inside the box or vice-versa, but permits optimisations across the entire box,
```
The regular and power workflows below assume that optimization happens before the assignment of boxes.

ihincks replied:
Although we want the boxes for twirling and mitigation, this RFC defines them more generally, as a core generic construct in qiskit. So the workflows below are specific to the use-case of twirling, but the construct itself needs to have this sort of generic interaction with the transpiler well-defined.

ewinston committed:
Merge remote-tracking branch 'SamFerracin/circuit-block' into box_dd

SamFerracin committed:
Merge pull request #3 from ewinston/box_dd

add section about dynamical decoupling with boxes

jakelishman commented:
I'll come back and fix up all the bits that were waiting for me while I was out tomorrow morning.

jakelishman committed:
Clarify basic `box` semantics

jakelishman committed:
Remove distracting discussion of `BackendV3` execution methods

These were hypothetical anyway, and should not be part of a new
`BackendV3`; there is no necessity, and it would be unacceptable API
whiplash for users anyway.

jakelishman commented:
I've now responded to Ian's review in #5.

jakelishman committed:
Update 0022-circuit-block.md

Co-authored-by: Ian Hincks <ian.hincks@gmail.com>

SamFerracin committed:
Merge pull request #5 from jakelishman/box/ian-review

Respond to Ian's review

jakelishman commented:
From my side, I'm happy to open this to a wider audience.
